File: Python/problem1232.py
from typing import List

class Solution:
    def checkStraightLine(self, coordinates: List[List[int]]) -> bool:
        # marks = [0] * (len(coordinates)-1)
        slopes = []
        for k in range(1, len(coordinates)):
            x0, y0 = coordinates[k-1][0], coordinates[k-1][1]
            x1, y1 = coordinates[k][0], coordinates[k][1]
            if x1 == x0:
                # marks[k-1] == 1
                slopes.append('inf')
            else:
                slopes.append((y1-y0)/(x1-x0))
            if len(slopes) > 1:
                if slopes[-1] != slopes[-2]:
                    return False
        return True

File: C++/problem1232.cpp
class Solution {
public:
    void printSet0(set<int> & x)
    {
        for (set<int>::iterator it = x.begin(); it != x.end(); ++it)
        {
            cout << *it << ' ';
        }
        cout << endl;
    }

    void printSet1(set<float> & x)
    {
        for (set<float>::iterator it = x.begin(); it != x.end(); ++it)
        {
            cout << *it << ' ';
        }
        cout << endl;
    }

    bool checkStraightLine(vector<vector<int>>& coordinates) {
        set<int> s;
        for (vector<vector<int>>::iterator it = coordinates.begin(); it < coordinates.end(); ++it)
        {
            s.insert((*it).front());
        }

        // printSet0(s);

        if (s.size() == 1)
        {
            return true;
        }

        set<float> t;
        for (vector<vector<int>>::iterator it = coordinates.begin() + 1; it < coordinates.end(); ++it)
        {
            if ((*(it)).front() == (*(it - 1)).front())
            {
                return false;
            }
            // cout << (float)(((*(it)).back() - (*(it - 1)).back()) / ((*(it)).front() - (*(it - 1)).front())) << endl;
            t.insert((float)((*(it)).back() - (*(it - 1)).back()) / (float)((*(it)).front() - (*(it - 1)).front()));
        }

        // printSet1(t);

        return t.size() == 1;
    }
};
